flict/flictlib/lic_comp.py

```diff
@@ -0,0 +1,217 @@
+            }
+
+        """
+        parsed_outbound = self.license_parser.parse_license([outbound])['license']
+        #assert(self.license_parser.is_license(parsed_outbound))
+
+        logging.debug(f"inbounds_outbound_check({outbound}, {expr})")
+        parsed = self.license_parser.parse_license(expr)['license']
+        logging.debug(f"inbounds_outbound_check: parsed:{self.license_parser.parse_license(expr)}")
+        logging.debug(f"inbounds_outbound_check: parsed:{parsed}")
+
+        compats = self._inbounds_outbound_check(parsed_outbound, parsed)
+
+        return compats
+
+    def _inbounds_outbound_check(self, outbound, expr):
+        logging.debug(f"_inbounds_outbound_check({outbound}, {expr})")
+        if self.license_parser.is_license(expr):
+            license = self.license_parser.license(expr)
+
+            outbound_aliased = self.alias.replace_aliases(self.license_parser.license(outbound))
+            license_aliased = self.alias.replace_aliases(license)
+
+            compat = self.compatibility.check_compat(outbound_aliased, license_aliased)
+
+            expr['license_aliased'] = license_aliased
+            expr['check'] = 'inbounds_outbound'
+            expr['outbound'] = outbound
+            expr['outbound_aliased'] = outbound_aliased
+            expr['allowed'] = self.license_parser.license_allowed(license)
+
+            expr[COMPATIBILITY_TAG] = compat[COMPATIBILITY_TAG]
+            return expr
+
+        elif self.license_parser.is_operator(expr):
+
+            compat_summary = None
+            allowed_summary = None
+            op = self.license_parser.operator(expr)
+            operands = self.license_parser.operands(expr)
+            for operand in operands:
+                logging.debug(f"Check operand: {operand}")
+
+                # get compatibility_tag between the operand and the outbound
+                # and calculate and store the summarized compatibility
+                compat = self._inbounds_outbound_check(outbound, operand)
+                compat_summary = self._update_compat(op, compat_summary, compat[COMPATIBILITY_TAG] == CompatibilityStatus.LICENSE_COMPATIBILITY_COMPATIBLE.value)
+
+                # are licenses allowed or denied for this expression
+                allowed = compat['allowed']
+                allowed_summary = self._update_allowed(op, allowed_summary, allowed)
+                operand['allowed'] = allowed
+                operand['allowed'] = allowed
+#                operand['compatibility_info'] = compat
+                operand[COMPATIBILITY_TAG] = compat[COMPATIBILITY_TAG]
+
+            # store outbound to make for easier reading of result
+            expr['outbound'] = outbound
+            #expr['compatibility_info'] = compat
+            expr[COMPATIBILITY_TAG] = compat_summary
+
+            expr['outbound'] = outbound
+            expr["allowed"] = allowed_summary
+            expr['check'] = 'inbounds_outbound'
```

Comment:
Lookup of keys is expensive. `expr` should be created inline instead of being expanded by key value after key value.

Reply:
Redacting my statement. I thought, this dict whould be newly created in here.